os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,264 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+	sf->byteswritten = last_data;
+	sf->entry.datalen = length;
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
+	*(uint16_t *)chainheader->used = SMARTFS_ERASEDSTATE_16BIT;
+	smartfs_sync_internal(fs, sf);
 
-	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
-		/* Read the next sector's header into our buffer */
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (FAR uint8_t *)fs->fs_rwbuffer);
 
-		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (uint8_t *)fs->fs_rwbuffer);
+	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			fdbg("Error reading sector %d header, ret : %d\n", readwrite.logsector, ret);
-			return ret;
+			fdbg("Unable to read sector %d, error = %d\n", readwrite.logsector, ret);
+			break;
+		}
+		nextsector = SMARTFS_NEXTSECTOR(chainheader);
+		ret = FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)readwrite.logsector);
+		if (ret < 0) {
+			fdbg("Unable to free sector %d to shrink file, error = %d\n", readwrite.logsector, ret);
+			break;
 		}
+		readwrite.logsector = nextsector;
+	}
 
-		/* Get the next chained sector */
+	return ret;
+}
 
-		sector = SMARTFS_NEXTSECTOR(header);
+/****************************************************************************
+ * Name: smartfs_extendfile
+ *
+ * Description:
+ *   Zero-extend the length of a regular file to 'length'.
+ *
+ ****************************************************************************/
 
-		/* If this is the 1st sector of the file, then just overwrite
-		 * the sector data with the erased state value.  The underlying
-		 * SMART block driver will detect this and release the old
-		 * sector and create a new one with the new (blank) data.
-		 */
+int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
+{
+	int ret;
+	size_t byteswritten;
+	ssize_t writelen;
+	uint16_t data_len;
+	char *buf = NULL;
 
-		if (nextsector == entry->firstsector) {
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	if (length == 0) {
+		return OK;
+	}
 
-			/* When we have a sector buffer in use, simply skip the first sector */
+	byteswritten = 0;
 
-			nextsector = sector;
-			continue;
+	smartfs_seek_internal(fs, sf, 0, SEEK_END);
+	length -= sf->entry.datalen;
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+	buf = (char *)kmm_malloc(data_len);
+	if (!buf) {
+		fdbg("Error allocating space for buffer\n");
+		return -ENOMEM;
+	}
+	memset(buf, '\0', data_len);
+	writelen = data_len;
 
-#else
+	while (length > 0) {
+		if (writelen > length) {
+			writelen = length;
+		}
+		if ((fs->fs_llformat.availbytes != sf->curroffset) && (writelen > (fs->fs_llformat.availbytes - sf->curroffset))) {
+			writelen = (fs->fs_llformat.availbytes - sf->curroffset);
+		}
```

Comment:
wirtelen will not changed to data_len forever if once it changed.

Reply:
Corrected

Reply:
you'd better make else after line 2075, to reduce one step of assembly level.